063-topic-replication-factor-change.md

```diff
@@ -0,0 +1,221 @@
+At most `max.active.user.tasks` tasks can be Active at the same time (5 by default), and only one task can be InExecution at any given time.
+
+#### Topic configuration endpoint
+
+The `topic_configuration` endpoint is used to request a replication factor change.
+Once accepted, the replication factor change is put into the Cruise Control's task queue and executed asynchronously.
+
+The two required parameters are `topic`, which is a regular expression to specify subject topics, and `replication_factor`, which is the target replication factor.
+In order to batch multiple changes in the same request, the Topic Operator uses a JSON payload containing topics grouped by their replication factor.
+
+```sh
+$ curl -vv -X POST -H "Content-Type: application/json" -d '{replication_factor:{topic_by_replication_factor:{2:topic1|topic3|topic4,3:topic2|topic5}}}' \
+  "http://localhost:9090/kafkacruisecontrol/topic_configuration?skip_rack_awareness_check=true&dryrun=false&json=true" | jq
+> POST /kafkacruisecontrol/topic_configuration?skip_rack_awareness_check=true&dryrun=false&json=true HTTP/1.1
+> Host: localhost:9090
+> User-Agent: curl/8.0.1
+> Accept: */*
+> 
+} [91 bytes data]
+< HTTP/1.1 200 OK
+< Date: Thu, 07 Dec 2023 14:58:55 GMT
+< Set-Cookie: JSESSIONID=node0lfkx183ijq4pjkkjahsegfmp30.node0; Path=/
+< Expires: Thu, 01 Jan 1970 00:00:00 GMT
+< User-Task-ID: 5344ca89-351f-4565-8d0f-9aade00e053d
+< Content-Type: application/json;charset=utf-8
+< Cruise-Control-Version: 2.5.77-SNAPSHOT
+< Cruise-Control-Commit_Id: 535ea07e8847990cc02dde1f1be99b387dbeaf5b
+< Content-Length: 17051
+< Server: Jetty(9.4.52.v20230823)
+< 
+{ [17321 bytes data]
+{
+  "summary": {
+    "numIntraBrokerReplicaMovements": 0,
+    "numReplicaMovements": 125,
+    "onDemandBalancednessScoreAfter": 100,
+    "intraBrokerDataToMoveMB": 0,
+    "monitoredPartitionsPercentage": 100,
+    "provisionRecommendation": "",
+    "excludedBrokersForReplicaMove": [],
+    "excludedBrokersForLeadership": [],
+    "provisionStatus": "RIGHT_SIZED",
+    "onDemandBalancednessScoreBefore": 100,
+    "recentWindows": 5,
+    "dataToMoveMB": 0,
+    "excludedTopics": [],
+    "numLeaderMovements": 0
+  }
+  ...
+```
+
+The `goals` parameter is the comma separated list of goals used to generate the configuration proposal.
+This is configurable for all endpoints in `Kafka.spec.cruiseControl.config` with `default.goals` property.
+
+The `replication_throttle` parameter is an upper bound on the bandwidth used to move replicas (bytes per second).
+This is configurable for all endpoints in `Kafka.spec.cruiseControl.config` with `default.replication.throttle` property.
+
+The `skip_rack_awareness_check` boolean parameter is undocumented and disabled by default.
+In case of a single-rack deployment, rack goals are not configured and the Topic Operator will automatically enable `skip_rack_awareness_check`.
```

Comment:
So, are these things known to Cruise Control? Or are they passed through the request by the Topic Operator?

Reply:
They are passed through the request by the Topic Operator. It's the Cluster Operator that sets STRIMZI_CRUISE_CONTROL_RACK_ENABLED env var, or the user in case of standalone deployment.

Reply:
Maybe worth adding here a sentence or two that these parameters will need to be passed to the TO since the list of env vars.

Reply:
Actually, the `STRIMZI_CRUISE_CONTROL_RACK_ENABLED` is mentioned int he environment variables. But nothing about goals or replication throttle.

Reply:
They are not configurable. I'm mentioning them in the "Topic configuration endpoint" section to say that they can be configured from Kafka.spec.cruiseControl.config.

Reply:
Ok. Maybe we should clarify it in the proposal as it is a bit confusing.